ddtrace/writer.py

```diff
@@ -188,45 +191,24 @@ def _apply_filters(self, traces):
         return traces
 
 
-class Q(object):
+class Q(Queue):
     """
     Q is a threadsafe queue that let's you pop everything at once and
     will randomly overwrite elements when it's over the max size.
     """
-    def __init__(self, max_size=1000):
-        self._things = []
-        self._lock = threading.Lock()
-        self._max_size = max_size
-        self._closed = False
-
-    def size(self):
-        with self._lock:
-            return len(self._things)
-
-    def close(self):
-        with self._lock:
-            self._closed = True
-
-    def closed(self):
-        with self._lock:
-            return self._closed
-
-    def add(self, thing):
-        with self._lock:
-            if self._closed:
-                return False
-
-            if len(self._things) < self._max_size or self._max_size <= 0:
-                self._things.append(thing)
-                return True
-            else:
-                idx = random.randrange(0, len(self._things))
-                self._things[idx] = thing
-
-    def pop(self):
-        with self._lock:
-            if not self._things:
-                return None
-            things = self._things
-            self._things = []
-            return things
+    def _init(self, maxsize):
+        self.queue = []
+
+    def put(self, item):
+        try:
+            # Cannot use super() here because Queue in Python2 is old style class
+            return Queue.put(self, item, False)
+        except Full:
+            # If the queue if full, replace a random item
+            idx = random.randrange(0, self.qsize())
+            self.queue[idx] = item
```

Comment:
Do we need to acquire lock here before modifying `self.queue`?

Reply:
Yeah you're right, we're doing two operations with `qsize` and insertion. Fixed.